datastore/prometheus.go

```diff
@@ -0,0 +1,197 @@
+}
+
+func NewPrometheusExporter(ctx context.Context) *PrometheusExporter {
+	exporter := &PrometheusExporter{
+		ctx:           ctx,
+		reg:           prometheus.NewRegistry(),
+		podCache:      newEventCache(),
+		svcCache:      newEventCache(),
+		reqChanBuffer: make(chan Request, 10000),
+	}
+
+	// Labels to consider using in the future:
+	// fromPod, fromNamespace, fromAcornProject, fromAcornApp, fromAcornAppNamespace, fromAcornContainer, fromAcornAccountId,
+	// toNamespace, toHost, toPort, toService, toAcornProject, toAcornAccountId
+
+	exporter.latencyHistogram = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "alaz",
+			Name:      "http_latency",
+			Buckets:   []float64{0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000},
+		},
+		[]string{"toPod", "toAcornApp", "toAcornContainer", "toAcornAppNamespace"},
+	)
+	exporter.reg.MustRegister(exporter.latencyHistogram)
+
+	exporter.statusCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "alaz",
+			Name:      "http_status",
+		},
+		[]string{"toPod", "toAcornApp", "toAcornContainer", "toAcornAppNamespace", "status"},
+	)
+	exporter.reg.MustRegister(exporter.statusCounter)
+
+	go exporter.handleReqs()
+
+	go func() {
+		http.Handle("/metricz", promhttp.HandlerFor(exporter.reg, promhttp.HandlerOpts{}))
+		go func() {
+			if err := http.ListenAndServe(":8080", nil); err != nil {
+				log.Logger.Error().Err(err).Msg("error while serving metrics")
+			}
+		}()
+		<-exporter.ctx.Done()
+		log.Logger.Info().Msg("Prometheus exporter stopped")
+	}()
+
+	return exporter
+}
+
+func (p *PrometheusExporter) handleReqs() {
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case req := <-p.reqChanBuffer:
+			p.handleReq(req)
+		}
+	}
+}
+
+func (p *PrometheusExporter) handleReq(req Request) {
+	if req.ToType == "pod" {
+		toPod, found := p.podCache.get(req.ToUID)
+		if found {
+			p.updateMetrics(toPod.(PodEvent), req)
+		}
+	}
+}
+
+func (p *PrometheusExporter) updateMetrics(toPod PodEvent, req Request) {
+	p.latencyHistogram.With(prometheus.Labels{
+		"toPod":               toPod.Name,
+		"toAcornApp":          toPod.Labels[appLabel],
+		"toAcornAppNamespace": toPod.Labels[appNamespaceLabel],
+		"toAcornContainer":    toPod.Labels[containerLabel],
+	}).Observe(float64(req.Latency) / float64(1000000)) // divide by 1 million to convert nanoseconds to milliseconds
+
+	p.statusCounter.With(prometheus.Labels{
+		"toPod":               toPod.Name,
+		"status":              strconv.Itoa(int(req.StatusCode)),
+		"toAcornApp":          toPod.Labels[appLabel],
+		"toAcornAppNamespace": toPod.Labels[appNamespaceLabel],
+		"toAcornContainer":    toPod.Labels[containerLabel],
+	}).Inc()
+}
+
+func (p *PrometheusExporter) PersistRequest(request Request) error {
+	p.reqChanBuffer <- request
+	return nil
+}
+
+func (p *PrometheusExporter) PersistPod(pod Pod, eventType string) error {
+	if eventType == "DELETE" {
```

Comment:
nit: is there a constant in the package that can be used here?

Reply:
There is, but it caused a dependency cycle if I imported it, because the package that declares them imports this one. I could move the constants to here, but I figured that wasn't really worth the effort.